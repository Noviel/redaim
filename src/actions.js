import { TRIGGER_UNKNOWN, TARGET_UNKNOWN, EXEC_UNKNOWN } from './constants';

export const createAction = 
  (trigger = TRIGGER_UNKNOWN) => 
  (target = TARGET_UNKNOWN) => 
  ({ exec = EXEC_UNKNOWN, payload = null, meta, ...rest } = {}) => 
({
  ...rest,
  type: trigger,
  payload,
  meta: {
    ...meta,
    target: target,
    exec
  }
});

export const aimFromAction = ({ type = '', payload = null, meta, ...rest } = {}) => {
  const [ 
    trigger = TRIGGER_UNKNOWN, 
    target = TARGET_UNKNOWN, 
    exec = EXEC_UNKNOWN 
  ] = type.split('/');

  return {
    ...rest,
    type: trigger || TRIGGER_UNKNOWN,
    payload,
    meta: {
      ...meta,
      target: target || TARGET_UNKNOWN,
      exec: exec || EXEC_UNKNOWN
    }
  };
};

export const wrapActions = fn => actions => {
  const finalActions = {};
  
  for (const a in actions) {
    const currAction = actions[a];
    if (typeof currAction === 'function') {
      finalActions[a] = (...args) => fn(currAction(...args));
    } else {
      finalActions[a] = (...args) => fn(currAction, ...args);
    }
  }

  return finalActions;
};

export const bindActions = actionCreatorWithTrigger => target => actions =>
  wrapActions(actionCreatorWithTrigger(target))(actions);
  
export const createActions = trigger => {
  const boundToTrigger = bindActions(createAction(trigger));

  return target => { 
    const boundToTarget = boundToTrigger(target);

    return actions => boundToTarget(actions);
  };
};

// reversed order of arguments
  
export const createActionR =
  ({ exec = EXEC_UNKNOWN, payload, meta, ...rest } = {}) =>
  (trigger = TRIGGER_UNKNOWN) =>
  (target = TARGET_UNKNOWN) =>
({
  ...rest,
  type: trigger,
  payload,
  meta: {
    ...meta,
    target: target,
    exec
  }
});

export const wrapActionsR = actions => fn => {
  const finalActions = {};
  
  for (const a in actions) {
    const currAction = actions[a];
    if (typeof currAction === 'function') {
      finalActions[a] = (...args) => fn(currAction(...args));
    } else {
      finalActions[a] = (...args) => fn(currAction, ...args);
    }
  }

  return finalActions;
};

export const bindActionsR = actions => {
  const wrappedActions = wrapActionsR(actions);

  return trigger => { 
    const actionAimer = createAction(trigger);

    return target => wrappedActions(actionAimer(target));
  };
};

export const createActionsR = actions => {
  const withAction = bindActionsR(actions);

  return trigger => {
    const withTrigger = withAction(trigger);

    return target => withTrigger(target);
  };
};
