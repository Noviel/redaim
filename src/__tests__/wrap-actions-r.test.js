import { wrapActionsR } from '../actions';

describe('wrapActionsR', () => {
  const actions = {
    'ACTION_ONE': {
      type: 'ACTION_ONE'
    },
    'ACTION_TWO': {
      type: 'ACTION_TWO'
    }
  };

  const identity = x => x;

  it('should return object with same keys as an input action', () => {
    const result = wrapActionsR(actions)(identity);
    const keys = Object.keys(result);

    expect(keys).toMatchSnapshot();
  });

  it('should return object with functions', () => {
    const result = wrapActionsR(actions)(identity);
    const keys = Object.keys(result);
    
    const toMatch = keys.reduce((acc, curr) => { 
      acc[curr] = typeof result[curr]; 
      return acc; 
    }, {});

    expect(toMatch).toMatchSnapshot();
  });

  it('should correctly apply function without arguments to each action', () => {
    const addConstant = action => ({ ...action, payload: 42 });
    const wrappedActions = wrapActionsR(actions)(addConstant);

    const wrappedActionOne = wrappedActions['ACTION_ONE']();
    const wrappedActionTwo = wrappedActions['ACTION_TWO']();

    expect({ wrappedActionOne, wrappedActionTwo }).toMatchSnapshot();
  });

  it('should correctly apply function with one argument to each action', () => {
    const addPayload = (action, payload) => ({ ...action, payload });
    const wrappedActions = wrapActionsR(actions)(addPayload);

    const wrappedActionOne = wrappedActions['ACTION_ONE'](10);
    const wrappedActionTwo = wrappedActions['ACTION_TWO']({ value: { inner: { innerValue: 100 }}});

    expect({ wrappedActionOne, wrappedActionTwo }).toMatchSnapshot();
  });

  it('should correctly apply function with multiple arguments to each action', () => {
    const addOneTwoToPayload = (action, one, two) => ({ ...action, payload: { one, two } });
    const wrappedActions = wrapActionsR(actions)(addOneTwoToPayload);

    const wrappedActionOne = wrappedActions['ACTION_ONE'](1, 'two');
    const wrappedActionTwo = wrappedActions['ACTION_TWO']({ value: { inner: { innerValue: 100 }}}, null);

    expect({ wrappedActionOne, wrappedActionTwo }).toMatchSnapshot();
  });

});
