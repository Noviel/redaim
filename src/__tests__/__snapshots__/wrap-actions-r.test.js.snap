// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapActionsR should correctly apply function with multiple arguments to each action 1`] = `
Object {
  "wrappedActionOne": Object {
    "payload": Object {
      "one": 1,
      "two": "two",
    },
    "type": "ACTION_ONE",
  },
  "wrappedActionTwo": Object {
    "payload": Object {
      "one": Object {
        "value": Object {
          "inner": Object {
            "innerValue": 100,
          },
        },
      },
      "two": null,
    },
    "type": "ACTION_TWO",
  },
}
`;

exports[`wrapActionsR should correctly apply function with one argument to each action 1`] = `
Object {
  "wrappedActionOne": Object {
    "payload": 10,
    "type": "ACTION_ONE",
  },
  "wrappedActionTwo": Object {
    "payload": Object {
      "value": Object {
        "inner": Object {
          "innerValue": 100,
        },
      },
    },
    "type": "ACTION_TWO",
  },
}
`;

exports[`wrapActionsR should correctly apply function without arguments to each action 1`] = `
Object {
  "wrappedActionOne": Object {
    "payload": 42,
    "type": "ACTION_ONE",
  },
  "wrappedActionTwo": Object {
    "payload": 42,
    "type": "ACTION_TWO",
  },
}
`;

exports[`wrapActionsR should return object with functions 1`] = `
Object {
  "ACTION_ONE": "function",
  "ACTION_TWO": "function",
}
`;

exports[`wrapActionsR should return object with same keys as an input action 1`] = `
Array [
  "ACTION_ONE",
  "ACTION_TWO",
]
`;
